contracts/near/src/governance.rs

```diff
@@ -0,0 +1,292 @@
+        Pyth,
+        PythExt,
+    },
+    near_sdk::{
+        borsh::{
+            self,
+            BorshDeserialize,
+            BorshSerialize,
+        },
+        env,
+        is_promise_success,
+        near_bindgen,
+        AccountId,
+        Balance,
+        Gas,
+        Promise,
+    },
+    std::io::Read,
+    structs::Chain as WormholeChain,
+    uint::byteorder::{
+        LittleEndian,
+        ReadBytesExt,
+    },
+};
+
+/// A `GovernanceAction` represents the different actions that can be voted on and executed by the
+/// governance system.
+#[derive(BorshDeserialize, BorshSerialize)]
+pub enum GovernanceAction {
+    ContractUpgrade([u8; 32]),
+    SetDataSources(Vec<Source>),
+    SetGovernanceSource(Source),
+    SetStalePriceThreshold(u64),
+    SetUpdateFee(Balance),
+}
+
+const GOVERNANCE_MAGIC: &[u8] = b"5054474d";
+const GOVERNANCE_MODULE: u8 = 0x01;
+
+impl GovernanceAction {
+    fn deserialize(data: &[u8]) -> Self {
+        let mut cursor = std::io::Cursor::new(data);
+        let magic = cursor.read_u32::<LittleEndian>().unwrap();
+        let module: u8 = cursor.read_u8().unwrap();
+        let action: u8 = cursor.read_u8().unwrap();
+        let target: u16 = cursor.read_u16::<LittleEndian>().unwrap();
+
+        assert!(magic == u32::from_le_bytes(GOVERNANCE_MAGIC.try_into().unwrap()));
+        assert!(module == GOVERNANCE_MODULE);
+        assert!(target == 0 || target == WormholeChain::Near as u16);
+
+        match action {
+            0 => {
+                let mut hash = [0u8; 32];
+                cursor.read_exact(&mut hash).unwrap();
+                Self::ContractUpgrade(hash)
+            }
+
+            1 => {
+                let mut sources = Vec::new();
+                let count = cursor.read_u32::<LittleEndian>().unwrap();
+                for _ in 0..count {
+                    let mut emitter = [0u8; 32];
+                    cursor.read_exact(&mut emitter).unwrap();
+                    let pyth_emitter_chain = Chain::from(WormholeChain::Solana);
+                    sources.push(Source {
+                        emitter,
+                        pyth_emitter_chain,
+                    });
+                }
+                Self::SetDataSources(sources)
+            }
+
+            2 => {
+                let mut emitter = [0u8; 32];
+                cursor.read_exact(&mut emitter).unwrap();
+                let pyth_emitter_chain = Chain(cursor.read_u16::<LittleEndian>().unwrap());
+                Self::SetGovernanceSource(Source {
+                    emitter,
+                    pyth_emitter_chain,
+                })
+            }
+
+            3 => {
+                let stale_price_threshold = cursor.read_u64::<LittleEndian>().unwrap();
+                Self::SetStalePriceThreshold(stale_price_threshold)
+            }
+
+            4 => {
+                let update_fee = cursor.read_u128::<LittleEndian>().unwrap();
+                Self::SetUpdateFee(update_fee)
```

Comment:
SetUpdateFee has to 64 bit numbers, value and exponent => fee = value * 10**expo